File: vscode/test/ime-provider-registry.test.js
const assert = require('node:assert/strict');
const test = require('node:test');

const { availableImeTypeFor, ImeType } = require('../out/core/types');
const { ImeProviderRegistry } = require('../out/ime/input/ImeProviderRegistry');

test('creates registered provider and rejects unsupported type', () => {
  const registry = new ImeProviderRegistry();
  const provider = { name: 'fake' };
  registry.register(ImeType.RIME, () => provider);

  assert.equal(registry.create({ type: ImeType.RIME }), provider);
  assert.throws(
    () => registry.create({ type: ImeType.SOGOU }),
    /IME provider not found/,
  );
});

test('unavailable or invalid config falls back to rime', () => {
  assert.equal(availableImeTypeFor(undefined), ImeType.RIME);
  assert.equal(availableImeTypeFor('invalid'), ImeType.RIME);
  assert.equal(availableImeTypeFor(ImeType.SOGOU), ImeType.RIME);
});

File: core/src/test/kotlin/com/auto_switch_ime/core/ime/input/ImeProviderRegistryTest.kt
package com.auto_switch_ime.core.ime.input

import com.auto_switch_ime.core.ImeConfig
import com.auto_switch_ime.core.ImeException
import com.auto_switch_ime.core.ImeProvider
import com.auto_switch_ime.core.ImeType
import com.auto_switch_ime.core.util.Logger
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Test

class ImeProviderRegistryTest {

    @Test
    fun `unavailable or invalid config falls back to rime`() {
        assertEquals(ImeType.RIME, ImeType.fromConfig(null))
        assertEquals(ImeType.RIME, ImeType.fromConfig("invalid"))
        assertEquals(ImeType.RIME, ImeType.fromConfig(ImeType.SOGOU.configValue))
    }

    @Test
    fun `creates registered provider and rejects unsupported type`() {
        val registry = ImeProviderRegistry()
        registry.register(ImeType.RIME) { _, _ -> FakeProvider }

        assertEquals(FakeProvider, registry.createProvider(ImeConfig(ImeType.RIME), FakeLogger))
        assertThrows(ImeException.ProviderNotFound::class.java) {
            registry.createProvider(ImeConfig(ImeType.SOGOU), FakeLogger)
        }
    }

    private object FakeProvider : ImeProvider {
        override val type = ImeType.RIME
        override val name = "fake"
        override fun start() = Unit
        override fun dispose() = Unit
    }

    private object FakeLogger : Logger {
        override fun debug(msg: String) = Unit
        override fun info(msg: String) = Unit
        override fun warn(msg: String, e: Throwable?) = Unit
        override fun error(msg: String, e: Throwable?) = Unit
    }
}
